enarx-keep-sev-shim/src/gdt.rs

```diff
@@ -0,0 +1,174 @@
+    eprintln!("init_gdt");
+
+    #[inline(always)]
+    #[allow(clippy::integer_arithmetic)]
+    fn stack_end(stack_start: VirtAddr, num_pages: usize) -> VirtAddr {
+        stack_start + num_pages * core::mem::size_of::<Page>()
+    }
+
+    unsafe {
+        TSS = Some({
+            let mut tss = TaskStateSegment::new();
+
+            // The stack for RING 0
+            tss.privilege_stack_table[0] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+
+            // The stacks for the exceptions and interrupts
+            tss.interrupt_stack_table[0] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[1] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[2] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[3] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[4] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[5] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss.interrupt_stack_table[6] = {
+                static mut STACK: [Page; STACK_NUM_PAGES] = [Page::zeroed(); STACK_NUM_PAGES];
+                stack_end(VirtAddr::from_ptr(&STACK), STACK_NUM_PAGES)
+            };
+            tss
+        });
+    }
+
+    unsafe {
+        GDT = Some({
+            let mut gdt = GlobalDescriptorTable::new();
+
+            // `syscall` loads segments from STAR MSR assuming a data_segment follows `kernel_code_segment`
+            // so the ordering is crucial here. Star::write() will panic otherwise later.
+            let code_selector = gdt.add_entry(Descriptor::kernel_code_segment());
+            let data_selector = gdt.add_entry(Descriptor::UserSegment(
+                (DescriptorFlags::USER_SEGMENT
+                    | DescriptorFlags::PRESENT
+                    | DescriptorFlags::WRITABLE
+                    | DescriptorFlags::LONG_MODE)
+                    .bits(),
```

Comment:
I wonder if upstream x86_64 would be interested in this as `kernel_data_segment()`

Reply:
true